File: src/test/samples/simple/src/test/java/io/vlingo/gradle/actortest/Test2ProtocolTest.java
package io.vlingo.gradle.actortest;

import static org.junit.Assert.assertEquals;

import java.util.Collections;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;

import io.vlingo.actors.Definition;
import io.vlingo.actors.testkit.TestActor;
import io.vlingo.actors.testkit.TestWorld;

public class Test2ProtocolTest {
    private TestWorld testWorld;

    @Test
    public void testDoOneThing() {
        final TestActor<Test2Protocol> test2 = testWorld.actorFor(new Definition(Test2ProtocolActor.class, Collections.emptyList()), Test2Protocol.class);

        test2.actor().doOneThing();
        assertEquals("doOneThing", Test2ProtocolActor.doOneThing);
    }

    @Test
    public void testDoAnotherThingUsing() {
        final TestActor<Test2Protocol> test2 = testWorld.actorFor(new Definition(Test2ProtocolActor.class, Collections.emptyList()), Test2Protocol.class);

        test2.actor().doAnotherThingUsing("hello", 1);
        assertEquals("doAnotherThingUsing: hello1", Test2ProtocolActor.doAnotherThingUsing);
    }

    @Test
    public void testSomethingRatherWonderful() {
        final TestActor<Test1Protocol> test1 = testWorld.actorFor(new Definition(Test1ProtocolActor.class, Collections.emptyList()), Test1Protocol.class);
        final TestActor<Test2Protocol> test2 = testWorld.actorFor(new Definition(Test2ProtocolActor.class, Collections.emptyList()), Test2Protocol.class);

        test2.actor().somethingRatherWonderful(test1.actor());
        assertEquals("somethingRatherWonderful: 1testingonetwothree", Test2ProtocolActor.somethingRatherWonderful);
    }

    @Before
    public void setUp() {
        testWorld = TestWorld.start("test-proxy");

        Test1ProtocolActor.doSomethingElseWithValue = 0;
        Test1ProtocolActor.doSomethingWith = null;

        Test2ProtocolActor.doAnotherThingUsing = null;
        Test2ProtocolActor.doOneThing = null;
        Test2ProtocolActor.somethingRatherWonderful = null;
    }

    @After
    public void tearDown() {
        testWorld.terminate();
    }
}

File: src/test/resources/protocols/test/kotlin/io/vlingo/gradle/actortest/Test2ProtocolTest.kt
package io.vlingo.gradle.actortest

import org.junit.Assert.assertEquals

import org.junit.After
import org.junit.Before
import org.junit.Test

import io.vlingo.actors.Definition
import io.vlingo.actors.testkit.TestActor
import io.vlingo.actors.testkit.TestWorld

class Test2ProtocolTest {
    private lateinit var testWorld: TestWorld

    @Test
    fun testDoOneThing() {
        val test2 = testWorld.actorFor(Test2Protocol::class.java, Definition(Test2ProtocolActor::class.java, emptyList()))

        test2.actor().doOneThing()
        assertEquals("doOneThing", Test2ProtocolActor.getDoOneThing())
    }

    @Test
    fun testDoAnotherThingUsing() {
        val test2 = testWorld.actorFor(Test2Protocol::class.java, Definition(Test2ProtocolActor::class.java, emptyList()))

        test2.actor().doAnotherThingUsing("hello", 1)
        assertEquals("doAnotherThingUsing: hello1", Test2ProtocolActor.getDoAnotherThingUsing())
    }

    @Test
    fun testSomethingRatherWonderful() {
        val test1 = testWorld.actorFor(Test1Protocol::class.java, Definition(Test1ProtocolActor::class.java, emptyList()))
        val test2 = testWorld.actorFor(Test2Protocol::class.java, Definition(Test2ProtocolActor::class.java, emptyList()))

        test2.actor().somethingRatherWonderful(test1.actor())
        assertEquals("somethingRatherWonderful: 1testingonetwothree", Test2ProtocolActor.getSomethingRatherWonderful())
    }

    @Before
    fun setUp() {
        testWorld = TestWorld.start("test-proxy")

        Test1ProtocolActor.setDoSomethingElseWithValue(0)
        Test1ProtocolActor.setDoSomethingWith(null)

        Test2ProtocolActor.setDoAnotherThingUsing(null)
        Test2ProtocolActor.setDoOneThing(null)
        Test2ProtocolActor.setSomethingRatherWonderful(null)
    }

    @After
    fun tearDown() {
        testWorld.terminate()
    }
}
